src/ui/Pager/Pager.ts

```diff
@@ -225,10 +246,25 @@ export class Pager extends Component {
   }
 
   private handleNoResults(data: INoResultsEventArgs) {
-    var lastValidPage = this.computePagerBoundary(data.results.totalCountFiltered, data.results.totalCount).lastResultPage;
-    if (this.currentPage > lastValidPage) {
-      this.ignoreNextQuerySuccess = true;
-      this.setPage(lastValidPage);
+    this.ignoreNextQuerySuccess = true;
+    let lastValidPage;
+    if (data.results.totalCount > 0) {
+      // First scenario : The index returned less results than expected (because of folding).
+      // Recalculate the last valid page, and change to that new page.
+      const possibleValidPage = this.computePagerBoundary(data.results.totalCountFiltered, data.results.totalCount).lastResultPage;
+      if (this.currentPage > possibleValidPage) {
+        lastValidPage = possibleValidPage;
+      }
+    } else if (this.currentPage > this.getMaxNumberOfPagesForCurrentResultsPerPage()) {
+      // Second scenario : Someone tried to access a non-valid page by the URL for example, which is  higher than the current possible with the number of
+      lastValidPage = this.getMaxNumberOfPagesForCurrentResultsPerPage();
+    }
+
+    // This needs to be deferred because we still want all the "querySuccess" callbacks the complete their execution
+    // before triggering/queuing the next query;
+    // if we cannot find a lastValidPage to go to, do nothing : this means it's a query that simply return nothing.
+    if (lastValidPage != null) {
+      Defer.defer(()=> this.setPage(lastValidPage));
```

Comment:
I'm not sure about deferring to execute after other callbacks. Can't you handle this directly in the `querySuccess` event?